doc/dfiasco.1
.\" $Id: dfiasco.1,v 1.3 2000/07/16 17:35:11 hafner Exp $
.TH dfiasco 1 "April, 2000" "FIASCO" "Fractal Image And Sequence COdec"

.SH NAME
.B  dfiasco
\- Decompress and display a FIASCO file

.SH SYNOPSIS
.B  dfiasco \fP[\fIoption\fP]... [\fIfilename\fP]...

.ad b
.hy 1
.SH DESCRIPTION
\|\fBdfiasco\fP\| decompresses the named FIASCO file(s), or the
standard input if no file is named, and displays the image or video in
a X11 window. Alternatively, if option \fB\-o\fP is specified, the
decompressed images are written to image files in raw pgm(5) or ppm(5)
format.

.SH OPTIONS
All option names may be abbreviated; for example, --output may be
written --outp or --ou. For all options an one letter short option
is provided. Mandatory or optional arguments to long options are
mandatory or optional for short options, too. Both short and long
options are case sensitive.

.TP
\fB\-o\fP[\fIname\fP], \fB\-\-output=\fP[\fIname\fP]
Write decompressed image to the named file \fIname\fP.ppm (color
image) or \fIname\fP.pgm, do not display the image in a window. If
\fIname\fP=- then produce the image file on the standard output. The
optional argument \fIname\fP can be omitted, then the input filename
is used as basename with the suffix .ppm or .pgm. In case of video
streams, the frames are stored in the files \fIname\fP.\fBN\fP.ppm
where \fBN\fP is the frame number (of the form 00..0 - 99..9); output
on the standard output is not possible with video streams.

If \fIname\fP is a relative path and the environment variable
\fBFIASCO_IMAGES\fP is a (colon-separated) list of directories, then
the output file(s) are written to the first (writable) directory of
this list. Otherwise, the current directory is used to store the
output file(s).

.TP
\fB\-z\fP, \fB\-\-fast\fP
Decompress images in the 4:2:0 format; i.e., each chroma channel is
decompressed to an image of halved width and height. Use this option
on slow machines when the desired frame rate is not achieved; the
output quality is only slightly decreased. 

.TP
\fB\-d\fP, \fB\-\-double\fP
Double the size of the X11 window both in width and height; no pixel
interpolation is used, each pixel is just replaced by four identical
pixels.
 
.TP
\fB\-p\fP, \fB\-\-panel\fP
Show a panel with play, stop, pause, record and exit buttons to
control the display of videos. When pressing the record button, all
frames are decompressed and stored in memory. The other buttons work
in the usual way.
 
.TP
\fB\-m\fP \fIN\fP, \fB\-\-magnify=\fIN\fP
Set magnification of the decompressed image. Positive values enlarge
and negative values reduce the image width and height by a factor of
2^|\fIN\fP|.

.TP
\fB\-s\fP \fIN\fP, \fB\-\-smoothing=\fIN\fP
Smooth decompressed image(s) along the partitioning borders by the
given amount \fIN\fP. \fIN\fP is 0 (minimum) to 100 (maximum). When
using this option, the smoothing value specified in the FIASCO file is
overridden.

.TP
\fB\-F\fP \fIN\fP, \fB\-\-framerate=\fIN\fP
Set number of frames per second to \fIN\fP. When using this option,
the frame rate specified in the FIASCO file is overridden.

.TP
\fB\-v\fP, \fB\-\-version
Print \|\fBdfiasco\fP\| version number, then exit.

.TP
\fB\-f\fP \fIname\fP, \fB\-\-config=\fIname\fP
Load parameter file \fIname\fP to initialize the options of
\|\fBdfiasco\fP\|. See file \fIsystem.fiascorc\fP for an example of the
syntax. Options of \|\fBdfiasco\fP\| are set by any of
the following methods (in the specified order):

1.) Global ressource file \fIsystem.fiascorc\fP

2.) $HOME/.fiascorc

3.) command line

4.) --config=\fIname\fP

.TP
\fB\-h\fP, \fB\-\-info
Print brief help, then exit.

.TP
\fB\-H\fP, \fB\-\-help
Print detailed help, then exit.


.SH EXAMPLES
.TP
dfiasco < foo.wfa
Decompress the FIASCO file "foo.wfa" and display the image in a window.

.TP
dfiasco -o foo1.wfa foo2.wfa
Decompress the FIASCO files "foo1.wfa" and "foo2.wfa" and write the
frames to the image files "foo1.wfa.ppm" and "foo2.wfa.ppm".

.TP
dfiasco -oimage foo1.wfa
Decompress the FIASCO file "foo1.wfa" and write all 15
frames to the image files "image.00.ppm", ... , "image.14.ppm".

.TP
dfiasco --fast --magnify=-1 --double video.wfa
Decompress the FIASCO file "video.wfa" and display the frames in a
window. The decompression speed is as fast as possible: the image is
decompressed (in 4:2:0 format) at a quarter of its original size; then
the image is enlarged again by pixel doubling. 

.RE

.SH FILES
.PD 0
.TP
.I system.fiascorc
The systemwide initialization file.
.TP
.I $HOME/.fiascorc
The personal initialization file.
.PD

.SH ENVIRONMENT
.PD 0
.TP
.B FIASCO_IMAGES
Save path for image files. Default is "./".
.TP
.B FIASCO_DATA
Search path for FIASCO files. Default is "./".
.PD 

.SH "SEE ALSO"
.br
.BR cfiasco (1)
.br

Ullrich Hafner, Juergen Albert, Stefan Frank, and Michael Unger.
\fBWeighted Finite Automata for Video Compression\fP, IEEE Journal on
Selected Areas In Communications, January 1998
.br
Ullrich Hafner. \fBLow Bit-Rate Image and Video Coding with Weighted
Finite Automata\fP, Ph.D. thesis, Mensch & Buch Verlag, ISBN
3-89820-002-7, October 1999.

.SH AUTHOR
Ullrich Hafner <hafner@bigfoot.de>


NEWS
user visible changes since FIASCO 1.2:
	* Fixed 24bpp rendering (required for any type of plugins)
	* Added GIMP plugin (read FIASCO images and videos)

user visible changes since FIASCO 1.1:
	* Fixed smoothing bug
	* Split option class into decoder and coder option class

user visible changes since FIASCO 1.0:
	* New title and comment attributes in FIASCO stream
	* Added GTK+ based analysis tool afiasco
	* Added binary tree visualization tool bfiasco
	* Added FIASCO file editor efiasco
	* Fixed some bugs:
		- Choose the right visual if the X11 screen offers
		  multiple  visuals.
		- Renamed decoder `fps' and 'smooth' attribute to
		  `framerate' and `smoothing' since
		  default values differ.
		- Don't crash when saving videos to disk.
		- Use user defined basis when specified.


codec/dfiasco.c
/*
 *  dfiasco.c:		Decoder public interface
 *
 *  Written by:		Ullrich Hafner
 *		
 *  This file is part of FIASCO ([F]ractal [I]mage [A]nd [S]equence [CO]dec)
 *  Copyright (C) 1994-2000 Ullrich Hafner <hafner@bigfoot.de>
 */
 
/*
 *  $Date: 2000/10/28 17:39:30 $
 *  $Author: hafner $
 *  $Revision: 5.7 $
 *  $State: Exp $
 */

#include "config.h"

#include "types.h"
#include "macros.h"
#include "error.h"

#include "dfiasco.h"
#include "wfa.h"
#include "read.h"
#include "misc.h"
#include "bit-io.h"
#include "decoder.h"
#include "options.h"
#include "wfalib.h"

/*****************************************************************************

				prototypes
  
*****************************************************************************/

static dfiasco_t *
cast_dfiasco (fiasco_decoder_t *dfiasco);
static void
free_dfiasco (dfiasco_t *dfiasco);
static dfiasco_t *
alloc_dfiasco (wfa_t *wfa, video_t *video, bitfile_t *input,
	       int enlarge_factor, int smoothing, format_e image_format);

/*****************************************************************************

				public code
  
*****************************************************************************/

fiasco_decoder_t *
fiasco_decoder_new (const char *filename, const fiasco_d_options_t *options)
{
   try
   {
      bitfile_t        	 *input;	/* pointer to WFA FIASCO stream */
      wfa_t            	 *wfa;		/* wfa structure */
      video_t          	 *video;	/* information about decoder state */
      const d_options_t  *dop;		/* decoder additional options */
      dfiasco_t	       	 *dfiasco;	/* decoder internal state */
      fiasco_decoder_t 	 *decoder;	/* public interface to decoder */
      fiasco_d_options_t *default_options = NULL;

      if (options)
      {
	 dop = cast_d_options ((fiasco_d_options_t *) options);
	 if (!dop)
	    return NULL;
      }
      else
      {
	 default_options = fiasco_d_options_new ();
	 dop 		 = cast_d_options (default_options);
      }
      
      wfa   = alloc_wfa (NO);
      video = alloc_video (NO);
      input = open_wfa (filename, wfa->wfainfo);
      read_basis (wfa->wfainfo->basis_name, wfa);

      decoder 	       	   = Calloc (1, sizeof (fiasco_decoder_t));
      decoder->delete  	   = fiasco_decoder_delete;
      decoder->write_frame = fiasco_decoder_write_frame;
      decoder->get_frame   = fiasco_decoder_get_frame;
      decoder->get_length  = fiasco_decoder_get_length;
      decoder->get_rate    = fiasco_decoder_get_rate;
      decoder->get_width   = fiasco_decoder_get_width;
      decoder->get_height  = fiasco_decoder_get_height;
      decoder->get_title   = fiasco_decoder_get_title;
      decoder->get_comment = fiasco_decoder_get_comment;
      decoder->is_color    = fiasco_decoder_is_color;

      decoder->private = dfiasco
		       = alloc_dfiasco (wfa, video, input,
					dop->magnification,
					dop->smoothing,
					dop->image_format);
   
      if (default_options)
	 fiasco_d_options_delete (default_options);
      if (dfiasco->enlarge_factor >= 0)
      {
	 int 	       n;
	 unsigned long pixels = wfa->wfainfo->width * wfa->wfainfo->height;

	 for (n = 1; n <= (int) dfiasco->enlarge_factor; n++)
	 {
	    if (pixels << (n << 1) > 2048 * 2048)
	    {
	       set_error (_("Magnifaction factor `%d' is too large. "
			    "Maximium value is %d."),
			  dfiasco->enlarge_factor, max (0, n - 1));
	       fiasco_decoder_delete (decoder);
	       return NULL;
	    }
	 }
      }
      else
      {
	 int n;

	 for (n = 0; n <= (int) - dfiasco->enlarge_factor; n++)
	 {
	    if (wfa->wfainfo->width >> n < 32
		|| wfa->wfainfo->height >> n < 32)
	    {
	       set_error (_("Magnifaction factor `%d' is too small. "
			    "Minimum value is %d."),
			  dfiasco->enlarge_factor, - max (0, n - 1));
	       fiasco_decoder_delete (decoder);
	       return NULL;
	    }
	 }
      }
      return (fiasco_decoder_t *) decoder;
   }
   catch
   {
      return NULL;
   }
}

int
fiasco_decoder_write_frame (fiasco_decoder_t *decoder,
			    const char *filename)
{
   dfiasco_t *dfiasco = cast_dfiasco (decoder);
   
   if (!dfiasco)
      return 0;
   else
   {
      try
      {
	 image_t *frame = get_next_frame (NO, dfiasco->enlarge_factor,
					  dfiasco->smoothing, NULL,
					  FORMAT_4_4_4, dfiasco->video, NULL,
					  dfiasco->wfa, dfiasco->input);
	 write_image (filename, frame);
      }
      catch
      {
	 return 0;
      }
      return 1;
   }
}

fiasco_image_t *
fiasco_decoder_get_frame (fiasco_decoder_t *decoder)
{
   dfiasco_t *dfiasco = cast_dfiasco (decoder);
   
   if (!dfiasco)
      return NULL;
   else
   {
      try
      {
	 fiasco_image_t *image = Calloc (1, sizeof (fiasco_image_t));
	 image_t 	*frame = get_next_frame (NO, dfiasco->enlarge_factor,
						 dfiasco->smoothing, NULL,
						 dfiasco->image_format,
						 dfiasco->video, NULL,
						 dfiasco->wfa, dfiasco->input);

	 frame->reference_count++;	/* for motion compensation */
	 image->private    = frame;
	 image->delete     = fiasco_image_delete;
	 image->get_width  = fiasco_image_get_width;
	 image->get_height = fiasco_image_get_height;
	 image->is_color   = fiasco_image_is_color;
	 
	 return image;
      }
      catch
      {
	 return NULL;
      }
   }
}

unsigned
fiasco_decoder_get_length (fiasco_decoder_t *decoder)
{
   dfiasco_t *dfiasco = cast_dfiasco (decoder);
   
   if (!dfiasco)
      return 0;
   else
      return dfiasco->wfa->wfainfo->frames;
}

unsigned
fiasco_decoder_get_rate (fiasco_decoder_t *decoder)
{
   dfiasco_t *dfiasco = cast_dfiasco (decoder);
   
   if (!dfiasco)
      return 0;
   else
      return dfiasco->wfa->wfainfo->fps;
}

unsigned
fiasco_decoder_get_width (fiasco_decoder_t *decoder)
{
   dfiasco_t *dfiasco = cast_dfiasco (decoder);

   if (!dfiasco)
      return 0;
   else
   {
      unsigned width;
      
      if (dfiasco->enlarge_factor >= 0)
	 width = dfiasco->wfa->wfainfo->width << dfiasco->enlarge_factor;
      else
	 width = dfiasco->wfa->wfainfo->width >> - dfiasco->enlarge_factor;
      
      return width & 1 ? width + 1 : width;
   }
}

unsigned
fiasco_decoder_get_height (fiasco_decoder_t *decoder)
{
   dfiasco_t *dfiasco = cast_dfiasco (decoder);

   if (!dfiasco)
      return 0;
   else
   {
      unsigned height;
      
      if (dfiasco->enlarge_factor >= 0)
	 height = dfiasco->wfa->wfainfo->height << dfiasco->enlarge_factor;
      else
	 height = dfiasco->wfa->wfainfo->height >> - dfiasco->enlarge_factor;

      return height & 1 ? height + 1 : height;
   }
}

const char *
fiasco_decoder_get_title (fiasco_decoder_t *decoder)
{
   dfiasco_t *dfiasco = cast_dfiasco (decoder);

   if (!dfiasco)
      return NULL;
   else
      return dfiasco->wfa->wfainfo->title;
}

const char *
fiasco_decoder_get_comment (fiasco_decoder_t *decoder)
{
   dfiasco_t *dfiasco = cast_dfiasco (decoder);

   if (!dfiasco)
      return NULL;
   else
      return dfiasco->wfa->wfainfo->comment;
}

int
fiasco_decoder_is_color (fiasco_decoder_t *decoder)
{
   dfiasco_t *dfiasco = cast_dfiasco (decoder);

   if (!dfiasco)
      return 0;
   else
      return dfiasco->wfa->wfainfo->color;
}

int
fiasco_decoder_delete (fiasco_decoder_t *decoder)
{
   dfiasco_t *dfiasco = cast_dfiasco (decoder);
   
   if (!dfiasco)
      return 1;
   
   try
   {
      free_wfa (dfiasco->wfa);
      free_video (dfiasco->video);
      close_bitfile (dfiasco->input);
      strcpy (dfiasco->id, " ");
      Free (decoder);
   }
   catch
   {
      return 0;
   }

   return 1;
}

/*****************************************************************************

				private code
  
*****************************************************************************/

static dfiasco_t *
alloc_dfiasco (wfa_t *wfa, video_t *video, bitfile_t *input,
	       int enlarge_factor, int smoothing, format_e image_format)
/*
 *  FIASCO decoder constructor:
 *  Initialize decoder structure.
 *
 *  Return value:
 *	pointer to the new decoder structure
 */
{
   dfiasco_t *dfiasco = Calloc (1, sizeof (dfiasco_t));

   strcpy (dfiasco->id, "DFIASCO");
   
   dfiasco->wfa 	   = wfa;
   dfiasco->video 	   = video;
   dfiasco->input 	   = input;
   dfiasco->enlarge_factor = enlarge_factor;
   dfiasco->smoothing  	   = smoothing;
   dfiasco->image_format   = image_format;
   
   return dfiasco;
}

static void
free_dfiasco (dfiasco_t *dfiasco)
/*
 *  FIASCO decoder destructor:
 *  Free memory of given 'decoder' struct.
 *
 *  No return value.
 *
 *  Side effects:
 *	'video' struct is discarded.
 */
{
   Free (dfiasco);
}

static dfiasco_t *
cast_dfiasco (fiasco_decoder_t *dfiasco)
/*
 *  Cast pointer `dfiasco' to type dfiasco_t.
 *  Check whether `dfiasco' is a valid object of type dfiasco_t.
 *
 *  Return value:
 *	pointer to dfiasco_t struct on success
 *      NULL otherwise
 */
{
   dfiasco_t *this = (dfiasco_t *) dfiasco->private;
   if (this)
   {
      if (!streq (this->id, "DFIASCO"))
      {
	 set_error (_("Parameter `dfiasco' doesn't match required type."));
	 return NULL;
      }
   }
   else
   {
      set_error (_("Parameter `%s' not defined (NULL)."), "dfiasco");
   }

   return this;
}
